### PR TITLE
Disallow having gittracks handle clustergittrackobjects

### PR DESCRIPTION
**Pull request by DanielMorsing**
This should ensure that we don't create any objects that might be garbage collected due to a invalid owner reference.

### CODE REVIEW

**Comment by DanielMorsing** [comment]
This PR is based off https://github.com/pusher/faros/pull/169, so look at that before reviewing this

File: pkg/controller/gittrack/gittrack_controller.go [hunk 1]
```diff
@@ -536,6 +527,9 @@ func (r *ReconcileGitTrack) ignoreObject(u *unstructured.Unstructured, owner far
 		} else if ownerIsClusterGittrack && r.clusterGitTrackMode == farosflags.CGTMExcludeNamespaced {
 			return true, "namespaced resources cannot be managed by ClusterGitTrack", nil
 		}
+	} else if ownerIsGittrack {
+		// cluster scoped object managed from namespaced gittrack. Disallow
+		return true, "cluster scoped object managed from gittrack are not allowed", nil
```
Comment (JoelSpeed) [inline pkg/controller/gittrack/gittrack_controller.go]: We should capitalise `GitTrack` to be consistent, I'm also tempted to reorder the phrasing slightly, WDYT?
```suggestion
		return true, "a GitTrack cannot manage a cluster scoped resource", nil
```

File: pkg/controller/gittrack/handler_test.go [hunk 2]
```diff
@@ -374,6 +394,16 @@ var _ = Describe("Handler Suite", func() {
 			})
 		}
 
+		var AssertNamespacedChildNameWithColon = func() {
+			BeforeEach(func() {
+				gto = testutils.ExampleGitTrackObject.DeepCopy()
+				gto.SetName("deployment-nginx-name-what-got-a-colon-in-it")
```
Comment (mthssdrbrg) [inline pkg/controller/gittrack/handler_test.go]:
```suggestion
				gto.SetName("deployment-nginx-name-that-got-a-colon-in-it")
```

File: pkg/controller/gittrack/gittrack_controller.go (hunk 1 above)
Comment (mthssdrbrg) [inline pkg/controller/gittrack/gittrack_controller.go]:
```suggestion
		return true, "cluster-scoped object managed from GitTrack are not allowed", nil
```

**Review by mthssdrbrg — COMMENTED** [review]
Looks good, just two minor comments, wdyt?

**Review by JoelSpeed — COMMENTED** [review]
One suggestion for an improvement in the error message, open to other ideas though, would like `GitTrack` capitalised at a minimum though

File: pkg/controller/gittrack/handler_test.go (hunk 2 above)
Comment (JoelSpeed) [reply pkg/controller/gittrack/handler_test.go]:
```suggestion
				gto.SetName("deployment-nginx-name-that-has-a-colon-in-it")
```
If we are going for better grammar 😉

Comment (DanielMorsing) [reply pkg/controller/gittrack/handler_test.go]: This was my attempt at being funny

The name is right for the test, do you still want me to go change it?

**Comment by DanielMorsing** [comment]
Went for Joel's changes in the end for the error message

**Review by JoelSpeed — COMMENTED** [review]
I'm not sure how I feel about the switch statement for the watchers setup.

One the one hand, it's an enum so it makes a lot of sense, on the other hand, the if statements also kinda worked and were perhaps easier to read 🤔 Any opinions @mthssdrbrg?

File: pkg/controller/gittrack/gittrack_controller.go [hunk 3]
```diff
@@ -113,7 +113,19 @@ func add(mgr manager.Manager, r reconcile.Reconciler, opts *reconcileGitTrackOpt
...
+		// do nothing
+	case farosflags.CGTMIncludeNamespaced:
+		err = c.Watch(&source.Kind{Type: &farosv1alpha1.GitTrackObject{}}, &handler.EnqueueRequestForOwner{
+			IsController: true,
+			OwnerType:    &farosv1alpha1.ClusterGitTrack{},
+		})
+		if err != nil {
+			return err
+		}
+		fallthrough
```
[the remark this answers is not in the document]
Comment (mthssdrbrg) [reply pkg/controller/gittrack/gittrack_controller.go]: Second having a comment explaining this, as I'm not sure I follow its purpose.

[the remark this answers is not in the document]
Comment (DanielMorsing) [reply pkg/controller/gittrack/gittrack_controller.go]: Going to restructure this

**Comment by DanielMorsing** [comment]
something's up here, the lines for the test failures don't match up with the code on this branch

**Comment by DanielMorsing** [comment]
The test failure is a data race, but it couldn't restore the stack, so not sure where it's happening. It's also in a part of the codebase I haven't touched at all. Going to retest and see if it's still happening

**Review by mthssdrbrg — APPROVED** [review]

**Pull request merged**